fix to accept cumulative_seqlens from TransformersKwargs in FA

Kurt232 opened the pull request:
The helper _flash_attention_forward now falls back to the keys cumulative_seqlens_q/k that may arrive inside TransformersKwargs when the explicit cu_seq_lens_q/k arguments are absent

A warning is raised on conflict, ensuring users notice any override

# What does this PR do?

I note that current transformer use [TransformersKwargs](https://github.com/huggingface/transformers/blob/4912d5b490a39d621b9d6f1b80df446d99e6239a/src/transformers/utils/generic.py#L856) as extra, but it will refuse the cumulative_seqlens_q and cumulative_seqlens_k args in [flash_attention func](https://github.com/huggingface/transformers/blob/4912d5b490a39d621b9d6f1b80df446d99e6239a/src/transformers/modeling_flash_attention_utils.py#L563), as unmatched argument name.

So I updated `_flash_attention_forward` to ensure compatibility with TransformersKwargs.

Fixes #40193 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

Cyrilvallez commented:
Yes, we don't really want this kind of workaround that clutter the code - @Kurt232 do you want to fix the typings where @vasqu mentionned instead? Otherwise we can do it 🤗
But TLDR, let's fix the root cause immediately, instead of first merging this and then reverting

Cyrilvallez commented:
Also, @vasqu can you make sure the current typing would be BC? If you changed it in your refactor, it may need to be update in `flash_attention_forward` instead of the `Kwargs` classes - otherwise it would plainly break BC for all downstream libs that used to pass them

Kurt232 commented:
@Cyrilvallez I'm willing to contribute this as mentioned, so I just need to create new PR to fix it?
@vasqu I want to double check that I just need to rename the `kwargs` dataclass to fit the FA function signature? e.g. cumulative_seqlens_q/k -> cu_seq_lens_q/k.

vasqu commented:
@Kurt232 Yes, that's correct

Kurt232 committed:
fix to the typings which are unmatched to FA function signature

cumulative_seqlens_q/k -> cu_seq_lens_q/k:
- in the FlashAttentionKwargs in modeling_flash_attention_utils
- in the TransformersKwargs in generic
- in the PagedAttentionArgs in continuous_batching

It is **BC**, because they are created in `ContinuousBatchProcessor.setup_static_tensors:L762`, used in `ContinuousBatchingManager._model_forward:L1233` and destroyed with `ContinuousBatchProcessor`

Kurt232 committed:
format changes by ruff

vasqu commented on src/transformers/integrations/flash_paged.py:
```diff
@@ -48,7 +48,7 @@ def paged_attention_forward(
         softcap: float. Anything > 0 activates softcapping attention.
     """
-    k, v = cache.update(k, v, module.layer_idx, cumulative_seqlens_k=cumulative_seqlens_k, **kwargs)
+    k, v = cache.update(k, v, module.layer_idx, cu_seq_lens_k=cu_seq_lens_k, **kwargs)
```
```suggestion
    k, v = cache.update(k, v, module.layer_idx, **kwargs)
```
I don't even think we need this as kwarg. See https://github.com/huggingface/transformers/blob/a9cd0b220b093aa7ea478fb9a40a324e5d1a80d6/src/transformers/generation/continuous_batching.py#L311-L326

Kurt232 replied:
I agree.

Kurt232 replied:
The `**kwargs` in signification is also unused, so we can remove both.
```suggestion
    k, v = cache.update(k, v, module.layer_idx)
```

vasqu replied:
I assume that read/write index might be hidden within the kwargs, so don't want to risk it there

vasqu reviewed (commented):
Can you add a :rotating_light: to the title, this is definitely breaking (API wise) for continuous batching (CB)

I'm not sure if CB is user-facing in this case so it might not be too bad but would check with @ArthurZucker

Kurt232 committed:
Update src/transformers/integrations/flash_paged.py

unused function arg in `PagedAttentionCache.update`

Co-authored-by: Anton Vlasjuk <73884904+vasqu@users.noreply.github.com>

ArthurZucker commented on src/transformers/generation/continuous_batching.py:
```diff
@@ -801,8 +801,8 @@ def get_model_kwargs(self) -> PagedAttentionArgs:
             "attention_mask": self.attention_mask,
-            "cumulative_seqlens_q": self.cumulative_seqlens_q,
-            "cumulative_seqlens_k": self.cumulative_seqlens_k,
+            "cu_seq_lens_q": self.cu_seq_lens_q,
+            "cu_seq_lens_k": self.cu_seq_lens_k,
```
happy to only change here

ArthurZucker commented on src/transformers/integrations/flash_paged.py:
```diff
@@ -35,9 +35,9 @@ def paged_attention_forward(
         k: (total_k, nheads_k, headdim), where total_k = total number of key tokens in the batch.  but if there is a block table it can be the full k
         v: (total_k, nheads_k, headdim), where total_k = total number of key tokens in the batch.  but if there is a block table it can be the full v
-        cumulative_seqlens_q: (batch_size + 1,), dtype torch.int32. The cumulative sequence lengths
+        cu_seq_lens_q: (batch_size + 1,), dtype torch.int32. The cumulative sequence lengths
            of the sequences in the batch, used to index into q.
```
my main issue with this naming is that is is no helpful for newbies, cu does not mean anything!

ArthurZucker commented on src/transformers/utils/generic.py:
```diff
@@ -880,8 +880,8 @@ class TransformersKwargs(TypedDict, total=False):
     output_router_logits: Optional[bool]
-    cumulative_seqlens_q: Optional["torch.LongTensor"]
-    cumulative_seqlens_k: Optional["torch.LongTensor"]
+    cu_seq_lens_q: Optional["torch.LongTensor"]
+    cu_seq_lens_k: Optional["torch.LongTensor"]
```
I don't mind this as in general we advertised that htese are unchanged and passed to fa2 directly

Kurt232 replied:
I purpose the question in #40193, as I was misled by the `TransformersKwargs`and the finaly flash_attn_interface missed the `cumulative_seqlens_q/k` args. 

Actually, we still use `cu_seq_lens_q/k` in the `Inputs` and a lot of models create `cu_seq_lens_q/k` in Attention module.

vasqu commented:
(we can remove the 🚨 as well then - only CB mightve been breaking)

Kurt232 committed:
revert continuous_batching signiture, which is more meaningful

Kurt232 commented on src/transformers/integrations/flash_paged.py:
```diff
@@ -48,7 +48,7 @@ def paged_attention_forward(
+    k, v = cache.update(k, v, module.layer_idx, **kwargs)
 
     sliding_window = (-1, -1) if not getattr(module, "sliding_window", False) else (module.sliding_window, 0)
     if implementation is not None:
```
I think `paged_attention_forward` should use `cu_seq_lens_q/k` instead of `cumulative_seqlens_q/k` to keep coherency with `flash_attn_varlen_func`.

ArthurZucker reviewed (approved):
Okay! I don't want to fight over this, its a small nit!
